akka-projection-r2dbc/src/main/scala/akka/projection/r2dbc/internal/R2dbcOffsetStore.scala

```diff
@@ -490,29 +506,38 @@ private[projection] class R2dbcOffsetStore(
 
   def validateAll[Envelope](envelopes: immutable.Seq[Envelope]): Future[immutable.Seq[(Envelope, Validation)]] = {
     import Validation._
-    envelopes
-      .foldLeft(Future.successful((getInflight(), Vector.empty[(Envelope, Validation)]))) { (acc, envelope) =>
-        acc.flatMap {
-          case (inflight, filteredEnvelopes) =>
-            createRecordWithOffset(envelope) match {
-              case Some(recordWithOffset) =>
-                validate(recordWithOffset, inflight).map {
-                  case Accepted =>
-                    (
-                      inflight.updated(recordWithOffset.record.pid, recordWithOffset.record.seqNr),
-                      filteredEnvelopes :+ (envelope -> Accepted))
-                  case rejected =>
-                    (inflight, filteredEnvelopes :+ (envelope -> rejected))
-                }
-              case None =>
-                Future.successful((inflight, filteredEnvelopes :+ (envelope -> Accepted)))
-            }
+    val result =
+      envelopes
+        .foldLeft(Future.successful((getInflight(), Vector.empty[(Envelope, Validation)]))) { (acc, envelope) =>
+          acc.flatMap {
+            case (inflight, filteredEnvelopes) =>
+              createRecordWithOffset(envelope) match {
+                case Some(recordWithOffset) =>
+                  validate(recordWithOffset, inflight).map {
+                    case Accepted =>
+                      (
+                        inflight.updated(recordWithOffset.record.pid, recordWithOffset.record.seqNr),
+                        filteredEnvelopes :+ (envelope -> Accepted))
+                    case rejected =>
+                      (inflight, filteredEnvelopes :+ (envelope -> rejected))
+                  }
+
+                case None =>
+                  Future.successful((inflight, filteredEnvelopes :+ (envelope -> Accepted)))
+              }
+          }
         }
-      }
-      .map {
-        case (_, filteredEnvelopes) =>
-          filteredEnvelopes
-      }
+        .map {
+          case (_, filteredEnvelopes) =>
+            filteredEnvelopes
+        }
+
+    if (validationObservers.nonEmpty)
+      result.foreach(_.foreach {
+        case (env, validation) => notifyValidationObserver(env, validation)
+      })
```

Comment:
Should it maybe rather be andThen:ed onto the returned future so the observer is a part of the backpressure rather than potentially queued up in parallel on the ec?

Reply:
good point, I was thinking that it shouldn't slow down the offset store, but it wouldn't be good if the observer is slower then the offset store

Reply:
changed in https://github.com/akka/akka-projection/pull/1014/commits/5f54fc5a672bab9dd7d7a26644b0052078315294